etna/models/deadline_ma.py

```diff
@@ -156,41 +155,85 @@ def forecast(self, df: pd.DataFrame, prediction_size: int) -> np.ndarray:
-        history_targets = history_targets.loc[history_timestamps >= context_beginning]
-        history_timestamps = history_timestamps.loc[history_timestamps >= context_beginning]
-        future_timestamps = df["timestamp"].iloc[-prediction_size:]
-
-        index = pd.date_range(start=context_beginning, end=future_timestamps.iloc[-1])
-        res = np.append(history_targets.values, np.zeros(prediction_size))
-        res = pd.DataFrame(res)
-        res.index = index
-        for i in range(len(history_targets), len(res)):
+        history = df_history["target"]
+        history = history[history.index >= context_beginning]
+        if np.any(history.isnull()):
+            raise ValueError("There are NaNs in a forecast context, forecast method required context to filled!")
+
+        index = pd.date_range(start=context_beginning, end=df.index[-1], freq=self._freq)
+        res = np.append(history.values, np.zeros(prediction_size))
+        res = pd.Series(res, index=index)
+        for i in range(len(history), len(res)):
             for w in range(1, self.window + 1):
                 if self.seasonality == SeasonalityMode.month:
                     prev_date = res.index[i] - pd.DateOffset(months=w)
                 elif self.seasonality == SeasonalityMode.year:
                     prev_date = res.index[i] - pd.DateOffset(years=w)
 
-                if prev_date <= history_timestamps.iloc[-1]:
-                    res.loc[index[i]] += history_targets.loc[history_timestamps == prev_date].values
-                else:
-                    res.loc[index[i]] += res.loc[prev_date].values
+                res.loc[index[i]] += res.loc[prev_date]
 
             res.loc[index[i]] = res.loc[index[i]] / self.window
 
         res = res.values.ravel()[-prediction_size:]
         return res
 
-    def predict(self, df: pd.DataFrame, prediction_size: int) -> pd.DataFrame:
-        raise NotImplementedError("Method predict isn't currently implemented!")
+    def predict(self, df: pd.DataFrame, prediction_size: int) -> np.ndarray:
+        """Compute predictions using true target data as context.
+
+        Parameters
+        ----------
+        df: pd.DataFrame
+            Used only for getting the horizon of forecast and timestamps.
+        prediction_size:
+            Number of last timestamps to leave after making prediction.
+            Previous timestamps will be used as a context for models that require it.
+
+        Returns
+        -------
+        :
+            Array with predictions.
+
+        Raises
+        ------
+        ValueError:
+            if context isn't big enough
+        ValueError:
+            if there are NaNs in a target column on timestamps that are required to make predictions
+        """
+        context_beginning = self._get_context_beginning(
+            df=df, prediction_size=prediction_size, seasonality=self.seasonality, window=self.window
+        )
+
+        df = df.set_index("timestamp")
```

Comment:
May be we can somehow rewrite it with _make_predictions method, as the only difference here is that in forecast we use predicted values and in predict the true values as context. E.g. in forecast case we will update context with predicted values and in predict we will not